### ISSUE TITLE
Child container is injecting root container into resolved instances

### ISSUE THREAD

**Issue by sharkBiscuit** [aurelia/dependency-injection#45]
Hi,

Here a minimal example of what I mean (in TypeScript..)

```typescript
let root = new Container();
let child = root.createChild();

@autoinject
class Test {
  constructor(container: Container) {
    console.log(container === root);  // true
    console.log(container === child); // false
  }
}

child.get(Test);
```

In that scenario, I would expect the child container to inject itself into the resolved instance and not the root. 

Cheers,
Andreas

**Comment by EisenbergEffect** [comment]
This is by design. If you do not manually register a type with the container first, auto-registration always occurs on the root container. You can control this with a registration decorator.

```
let root = new Container();
let child = root.createChild();

@singleton(true)
@autoinject
class Test {
  constructor(container: Container) {
    console.log(container === root);  // false
    console.log(container === child); // true
  }
}

child.get(Test);
```
Passing `true` indicates that when auto-registering the type, it should be registered with the child container.

**Comment by EisenbergEffect** [comment]
To manually register, you would do this:

```
let root = new Container();
let child = root.createChild();

@autoinject
class Test {
  constructor(container: Container) {
    console.log(container === root);  // false
    console.log(container === child); // true
  }
}

child.registerSingleton(Test);
child.get(Test);
```

**Comment by sharkBiscuit** [comment]
Thank you for the quick response!

child.registerSingleton(Test) => works

@singleton(true) => does not.

I quickly checked it and the annotation doesn't seem to have an effect. 
The registration into the child works like a charm.

**Comment by EisenbergEffect** [comment]
Ok, let me look into it. That should have worked. There may be a bug there.

**Comment by EisenbergEffect** [comment]
I apologize. I told you wrong. The singleton decorator will work...but only if the type truly needs to be registered. In this case, it already exists in the parent container, so it doesn't register it. It just resolves from the parent. The two ways to accomplish this are to manually register in the child (which is actually what the framework itself does in certain scenarios ie the router and the view factory) or you can create a custom Resolver. This resolver and be used in the inject decorator in place of the type to direct the container that it wants to resolve the dependency, but directly from the leaf container, and never from the parent.

```
export class Scoped extends Resolver {
  constructor(key:any){
    super();
    this.key = key;
  }

  get(container:Container):any[]{
    if(!container.hasHandler(this.key){
      container.registerSingleton(key);
    }

    return container.get(this.key);
  }
```

And use it like this:

```
let root = new Container();
let child = root.createChild();

@autoinject
class Test {
  constructor(container: Container) {
    console.log(container === root);  // false
    console.log(container === child); // true
  }
}

child.get(new Scoped(Test));
```

Not that if you had a class that wanted Test as a dependency, but alway scoped to the child container, you would do this:

```
@inject(new Scoped(Test))
class Another {
  constructor(test: Test) {

  }
}
```

**Comment by thomas-darling** [comment]
@EisenbergEffect would it be possible to create an `@scoped` decorator that can be applied to the class, that is being instantiated, just like `@singleton` and `@transient`?

We have a case here, where a service needs access to the child container of the component into which it is being injected, and the solution you presented above would be quite impractical, as it means we can't use `@autoinject` in a lot of our view models.

For context, what we're facing here, is some rather interesting challenges related to tracking of user interactions, where we need to walk up the container tree, to figure out where we are in the app - basically a "component path" from the root of the app to the current component.
Suggestions for how we could otherwise achieve this would be very welcome - but either way, I think a `@scoped` lifetime decorator would be *really* handy for both this and other scenarios.

**Comment by EisenbergEffect** [comment]
Give a shot at the implementation and then let's review it and see.

**Comment by thomas-darling** [comment]
Hmm, I just looked into this a little deeper, and I'm seeing some unexpected behavior...

```typescript
import {Registration, Resolver, Container, registration, autoinject} from "aurelia-framework";

export function scoped(): any
{
    return registration(new ScopedRegistration());
}

export class ScopedRegistration implements Registration
{
    public registerResolver(container: Container, key: any, fn: Function): Resolver
    {
        console.log("registerResolver", key.name, container.parent != null);
        return container.registerSingleton(key, fn);
    }
}

@autoinject
@scoped()
export class SomeService
{
    public constructor(container: Container)
    {
        console.log("SomeService", container.parent != null);
    }
}

@autoinject
@scoped()
export class SomeComponentViewModel
{
    public constructor(container: Container)
    {
        console.log("SomeComponentViewModel", container.parent != null);
    }
}
```
This results in
```
registerResolver  SomeComponentViewModel true
registerResolver  SomeService false
SomeService false
SomeComponentViewModel true
```
When applied to a component view model, the registration correctly happens in the child container. But when applied to a service injected into the view model, the registration always happens in the root container. Is this really the intended behavior?

Of course, it makes perfect sense that we generally want services registered in the root container by default, but here it looks like the `ScopedRegistration` does not even get a chance to change that - it never even sees the child container when registering a service - only when registering a view model.

Also, I could not make the code you posted a while ago work - it seems the `get` method on the `Resolver` is never called - not sure what to make of that?

**Comment by thomas-darling** [comment]
Ok, so I got a `Scoped` resolver working - it's essentially the code posted back in August, updated to match the current version of the libraries - maybe this could be added to the framework itself?

```typescript
import {Container, Resolver, resolver, getDecoratorDependencies} from "aurelia-framework";

/**
 * Decorator: Specifies the dependency should be scoped to the current container.
 */
export function scoped(keyValue: any)
{
    return function(target, key, index)
    {
        let params = getDecoratorDependencies(target, "scoped");
        params[index] = Scoped.of(keyValue);
    };
}

/**
 * Used to allow functions/classes to specify scoped resolution logic.
 */
@resolver()
export class Scoped implements Resolver
{
    private _key: any;

    /**
     * Creates an instance of the Scoped class.
     * @param key The key to resolve.
     */
    public constructor(key: any)
    {
        this._key = key;
    }

    /**
     * Creates a Scoped Resolver for the supplied key.
     * @param key The key to resolve.
     * @return Returns an instance of Scoped for the key.
     */
    public static of(key: any): Scoped
    {
        return new Scoped(key);
    }

    /**
     * Called by the container to load the dependency from the current container
     * @param container The container to resolve from.
     * @return Returns an instance scoped to the current container.
     */
    public get(container: Container): any
    {
        if (!container.hasResolver(this._key, false))
        {
            container.registerSingleton(this._key);
        }

        return container.get(this._key);
    }
}
```

This can then be used in two ways:

```typescript
@inject(Scoped.of(SomeService))
class SomeComponentViewModel
{
    constructor(someService: SomeService)
    {
    }
}
```
```typescript
@autoinject
class SomeComponentViewModel
{
    constructor(@scoped(SomeService) someService: SomeService)
    {
    }
}
```

**Comment by thomas-darling** [comment]
Anyway, despite getting the above code working, I'm still concerned that it's seemingly impossible to have the class of a dependency declare *itself* as being scoped to the current container. Having to remember to *always* apply the `@scoped` resolver to a constructor parameter is far from ideal - it's pretty much guaranteed to be forgotten at some point.

I also think it's a bit of a problem that the behavior of `@singleton(true)` is inconsistent - it only works on view models, and not on their dependencies - it's a highly unexpected behavior. I would expect a dependency with this class decorator applied, to be registered in the same container as the view model depending on it - assuming, of course, it is the first time that dependency is being resolved.

Thoughts?

**Comment by thomas-darling** [comment]
For future reference, the above mentioned concerns all seem to be resolved, as of commit https://github.com/aurelia/dependency-injection/commit/98739f2, released in `aurelia-dependency-injection 1.2.1`.

Decorating a service with `@singleton(true)` now causes the service to always be registered in, and resolved from, the same child container as the view model - exactly as one would expect.
